web3/_utils/batching.py

```diff
@@ -0,0 +1,171 @@
+from copy import (
+    copy,
+)
+from types import (
+    TracebackType,
+)
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Coroutine,
+    Dict,
+    Generic,
+    List,
+    Sequence,
+    Tuple,
+    Type,
+    Union,
+    cast,
+)
+
+from web3._utils.compat import (
+    Self,
+)
+from web3.contract.async_contract import (
+    AsyncContractFunction,
+)
+from web3.contract.contract import (
+    ContractFunction,
+)
+from web3.exceptions import (
+    Web3ValueError,
+)
+from web3.types import (
+    TFunc,
+    TReturn,
+)
+
+if TYPE_CHECKING:
+    from web3 import (  # noqa: F401
+        AsyncWeb3,
+        Web3,
+    )
+    from web3.method import (  # noqa: F401
+        Method,
+    )
+    from web3.providers import (  # noqa: F401
+        PersistentConnectionProvider,
+    )
+    from web3.providers.async_base import (  # noqa: F401
+        AsyncJSONBaseProvider,
+    )
+    from web3.types import (  # noqa: F401
+        RPCEndpoint,
+        RPCResponse,
+    )
+
+BatchRequestInformation = Tuple[Tuple["RPCEndpoint", Any], Sequence[Any]]
+RPC_METHODS_UNSUPPORTED_DURING_BATCH = {
+    "eth_subscribe",
+    "eth_unsubscribe",
+    "eth_sendRawTransaction",
+    "eth_sendTransaction",
+    "eth_signTransaction",
+    "eth_sign",
+    "eth_signTypedData",
+}
+
+
+class BatchRequestContextManager(Generic[TFunc]):
```

Comment:
Now that it's no longer just a context manager, e.g.:

```py
batch = w3.batch_requests()
batch.add(...)
batch.add(...)
result1, result2 = batch.execute()
```

... this class should probably be renamed to a more general name like `RequestBatcher`. We already have `RequestManager` and `RequestProcessor` classes. `RequestBatcher` seems like a fitting enough name.